From a90d5d80672d8a5414771b4b4701e514b29dde60 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Mon, 23 Jun 2014 21:50:03 -0400
Subject: [fs] nfsd: Don't return NFS4ERR_STALE_STATEID for NFSv4.1+

Message-id: <1403560314-32525-62-git-send-email-bfields@redhat.com>
Patchwork-id: 84530
O-Subject: [Patch RHEL7 061/172] nfsd: Don't return NFS4ERR_STALE_STATEID for NFSv4.1+
Bugzilla: 1111719
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

RFC5661 obsoletes NFS4ERR_STALE_STATEID in favour of NFS4ERR_BAD_STATEID.

Note that because nfsd encodes the clientid boot time in the stateid, we
can hit this error case in certain scenarios where the Linux client
state management thread exits early, before it has finished recovering
all state.

Reported-by: Idan Kedar <idank@primarydata.com>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: a8a7c6776f8d74780348bef639581421d85a4376
Bugzilla: 1111719
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 451638b..3ba6597 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -3627,8 +3627,11 @@ static __be32 nfsd4_lookup_stateid(stateid_t *stateid, unsigned char typemask,
   return nfserr_bad_stateid;
  status = lookup_clientid(&stateid->si_opaque.so_clid, sessions,
        nn, &cl);
- if (status == nfserr_stale_clientid)
+ if (status == nfserr_stale_clientid) {
+  if (sessions)
+   return nfserr_bad_stateid;
   return nfserr_stale_stateid;
+ }
  if (status)
   return status;
  *s = find_stateid_by_type(cl, stateid, typemask);
-- 
1.7.1